src/util/address.rs

```diff
@@ -501,7 +501,9 @@ impl ObjectReference {
     #[inline(always)]
     pub fn to_address<VM: VMBinding>(self) -> Address {
         use crate::vm::ObjectModel;
-        VM::VMObjectModel::ref_to_address(self)
+        let to_address = VM::VMObjectModel::ref_to_address(self);
+        debug_assert!(!VM::VMObjectModel::UNIFIED_OBJECT_REFERENCE_ADDRESS || to_address == self.to_raw_address(), "The binding claims unified object reference address, but for object reference {}, ref_to_address() returns {}", self, to_address);
```

Comment:
Hmmm... I think `ObjectReference::to_address` should be a synonym of `VM::VMObjectModel::ref_to_address`.  But since it is the VM that implements `VM::VMObjectModel::ref_to_address`, we can't offload this assertion to the binding.  This makes `ObjectReference::to_address` execute one extra assertion (when debugging, of course), and it feels odd.

I don't know if any kind of code instrumenting tool (aspect-oriented programming, design by contract, post conditions, etc.) can help inserting this assertion after every implementation of `VM::VMObjectModel::ref_to_address`.  If there is no such tool, I am OK with this.

Reply:
Thanks for pointing this out. It is a general goal for us to assert that a binding implements those traits correctly (see https://github.com/mmtk/mmtk-core/issues/480). However, I don't know if there is a better way to achieve this, other than just asserting at where we use those traits. If there is any good instrumentation tool for Rust, it is worth a try, as it can help us solve this general issue.

However, I wouldn't think that (debug) assertions change the semantics or the behavior of the method. They just help ensure some invariants (which are already part of the semantics) are not broken. Thus I don't think adding assertions changes the fact that `ObjectReference::to_address` is a synonym of `VM::VMObjectModel::ref_to_address`.